framework/framework.go

```diff
@@ -175,10 +163,40 @@ func (f *Framework) NewCacheResourceEventHandler() *cache.ResourceEventHandlerFu
 // UpdateFunc only redirects to AddFunc and only dispatches the new custom
 // object received.
 func (f *Framework) UpdateFunc(oldObj, newObj interface{}) {
-	f.AddFunc(newObj)
+	obj := newObj
+
+	// AddFunc/DeleteFunc/UpdateFunc is synchornized to make sure only one
+	// of them is executed at a time. AddFunc/DeleteFunc/UpdateFunc is not
+	// thread safe. This is important because the source of truth for an
+	// operator are the reconciled resources. In case we would run the
+	// operator logic in parallel, we would run into race conditions.
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
+
+	ctx := context.Background()
+	ctx = canceledcontext.NewContext(ctx, make(chan struct{}))
+
+	if f.initializer != nil {
+		var err error
+		ctx, err = f.initializer(ctx, obj)
+		if err != nil {
+			f.logger.Log("error", fmt.Sprintf("%#v", err), "event", "create")
+			return
+		}
+	}
+
+	f.logger.Log("action", "start", "component", "operatorkit", "function", "ProcessUpdate")
+
+	err := ProcessUpdate(ctx, obj, f.resources)
```

Comment:
Nitpick, this line and the next could be merged

Reply:
Yeah, and I hate this. :D